threadpoolctl.py

```diff
@@ -60,56 +67,322 @@ class _dl_phdr_info(ctypes.Structure):
-        "filename_prefixes": ("libopenblas", "libblas"),
-        "check_symbols": ("openblas_get_num_threads", "openblas_get_num_threads64_"),
-    },
-    "MKLController": {
-        "user_api": "blas",
-        "internal_api": "mkl",
-        "filename_prefixes": ("libmkl_rt", "mkl_rt", "libblas"),
-        "check_symbols": ("MKL_Get_Max_Threads",),
-    },
-    "BLISController": {
-        "user_api": "blas",
-        "internal_api": "blis",
-        "filename_prefixes": ("libblis", "libblas"),
-        "check_symbols": ("bli_thread_get_num_threads",),
-    },
-}
+class LibController(ABC):
+    """Abstract base class for the individual library controllers
+
+    A library controller must expose the following class attributes:
+        - user_api : str
+            Usually the name of the library or generic specification the library
+            implements.
+        - internal_api : str
+            Usually the name of the library, or concrete implementation of some
+            specification.
+        - filename_prefixes : tuple
+            Possible prefixes of the shared library's filename that allow to
+            identify the library. e.g. "libopenblas" for libopenblas.so.
+
+    A library contoller must implement the following methods: `get_num_threads`,
+    `set_num_threads` and `get_version`.
+
+    The following information will be exposed in the info dict:
+      - user_api : user API.
+      - internal_api : internal API.
+      - num_threads : the current thread limit.
+      - prefix : prefix of the shared library's filename.
+      - filepath : path to the loaded shared library.
+      - version : version of the library (if available).
+
+    In addition, each library controller may expose internal API specific entries. They
+    must be set as attributes in the `set_additional_attributes method`.
+    """
+
+    @final
+    def __init__(self, *, filepath=None, prefix=None):
+        self.prefix = prefix
+        self.filepath = filepath
+        self._dynlib = ctypes.CDLL(filepath, mode=_RTLD_NOLOAD)
+        self.version = self.get_version()
+        self.set_additional_attributes()
+
+    @final
+    def info(self):
+        """Return relevant info wrapped in a dict"""
+        all_attrs = {
+            "user_api": self.user_api,
+            "internal_api": self.internal_api,
+            "num_threads": self.num_threads,
+            **vars(self),
+        }
+        return {k: v for k, v in all_attrs.items() if not k.startswith("_")}
+
+    def set_additional_attributes(self):
+        """Set additional attributes meant to be exposed in the info dict"""
+
+    @property
+    @abstractmethod
+    def user_api(self):
+        """User API of the library"""
+
+    @property
+    @abstractmethod
+    def internal_api(self):
+        """Internal API of the library"""
+
+    @property
+    @abstractmethod
+    def filename_prefixes(self):
+        """Possible prefixes of the library's filename"""
+
+    @property
+    @final
```

Comment:
I did not know this one but this looks super useful!

Reply:
I learned about it while making this PR :)
It's from typing though, not from abc, meaning it's only used by type checkers but there's no runtime error if you still override a method decorated as final.